zigbee: Switch to IEEE 802.15.4 radio API

maciekfabia opened the pull request:
Switch Zigbee subsys to IEEE 802.15.4 radio API instead of using radio driver API directly.

It depends on:

- [x] https://github.com/nrfconnect/sdk-zephyr/pull/357

rlubos reviewed (commented):
Partial PR, halfway through.

rlubos commented on subsys/zigbee/osif/zb_nrf_transceiver.c:
```diff
@@ -228,26 +283,32 @@ zb_bool_t zb_trans_is_active(void)
 {
+	struct net_buf frag = {
+		.frags = NULL,
+		.b = { .data = &tx_buf[1],
```
Let's move the initializer to a new line, i. e.:
```
		.b = {
			 .data = &tx_buf[1],
			...
		}
```

maciekfabia replied:
fixed

rlubos commented on subsys/zigbee/osif/zb_nrf_transceiver.c:
```diff
@@ -228,26 +283,32 @@ zb_bool_t zb_trans_is_active(void)
-#endif
+		state_cache.radio_state = RADIO_802154_STATE_TRANSMIT;
+		err = radio_api->tx(radio_dev,
+			IEEE802154_TX_MODE_CSMA_CA,
```
nit: Alignment

rlubos commented on subsys/zigbee/osif/zb_nrf_transceiver.c:
```diff
@@ -257,332 +318,250 @@ zb_bool_t zb_trans_transmit(zb_uint8_t wait_type, zb_time_t tx_at,
-			tx_buf, ZB_FALSE);
+		state_cache.radio_state = RADIO_802154_STATE_TRANSMIT;
+		err = radio_api->tx(radio_dev,
+			IEEE802154_TX_MODE_DIRECT,
```
nit: Alignment

rlubos commented on subsys/zigbee/osif/zb_nrf_transceiver.c:
```diff
@@ -228,26 +283,32 @@ zb_bool_t zb_trans_is_active(void)
+		state_cache.radio_state = RADIO_802154_STATE_TRANSMIT;
+		err = radio_api->tx(radio_dev,
+			IEEE802154_TX_MODE_CSMA_CA,
+			NULL,
```
Have you verified that it's ok to pass a NULL `net_pkt`?

maciekfabia replied:
Yes, `pkt` is not used inside tx function. The packet that will be sent is provided via `frag` parameter. 
Radio API header comments about tx function kind of silimiar way too:
`/** Transmit a packet fragment */`

MarekPorwisz commented on subsys/zigbee/Kconfig:
```diff
@@ -18,6 +19,7 @@ menuconfig ZIGBEE
 	imply FLASH_PAGE_LAYOUT
 	imply FLASH_MAP
 	imply MPU_ALLOW_FLASH_WRITE
+	imply NETWORKING
```
Why do we imply networking? If zigbee requires networking to work it should be "select" in case you want it to enable when enabling zigbee or "depends on" if you want to be able to enable zigbee only when networking enabled (note depends on is preferred).  If it is still possible to work with zigbee with networking disable this seems to be ok. @rlubos what do you think?

maciekfabia replied:
Zigbee won't work without `NETWORKING` so it's a necessary component. The reason I used `IMPLY` is because `NETWORKING` is a visible symbol in menuconfig i.e. it has prompt, some docs say one should not use `SELECT` for visible symbols.
However @MarekPorwisz's argument also seems valid. @rlubos what is your opinion?

rlubos replied:
I think `select` is fine for this particular case, given that `NETWORKING` is a hard-dependency for zigbee now. If you prefer `depends on`, it fine as well, but keep in mind this will require to add `CONFIG_NETWORKING=y` to every single sample.

maciekfabia replied:
Changed to `SELECT`. Also, `NRF_802154_RADIO_DRIVER` removed as it's selected by `IEEE802154_NRF5`

maciekfabia commented on samples/zigbee/light_bulb/prj.conf:
```diff
@@ -39,3 +39,10 @@ CONFIG_CRYPTO_NRF_ECB=y
 CONFIG_ENTROPY_CC310=n
+
+#Networking
+CONFIG_NET_IPV6_MLD=n
```
When I disable `NET_IPV6` but `NET_NATIVE` is still enabled, I get error during net_pkt.c compilation:
`#error "Either IPv6 or IPv4 needs to be selected."`

When I disable both I get a lot of undefined references during linking to `net_pkt_*` stuff like `net_pkt_alloc_with_buffer`, `net_pkt_write`. These missing functions are needed in `ieee802154_nrf5.c`, `net_core.c`.

rlubos replied:
That's unfortunate, we end up building parts of the net stack that have no use for Zigbee. Not sure if I have a good solution for this (at least w/o further modifications in Zephyr).

There's a special Kconfig called [NET_RAW_MODE](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/CMakeLists.txt#L10) that disables literally everything in the net stack, beside `net_pkt`. Perhaps it could be used for Zigbee, but this config disables RX/TX queues as well (TX is not an issue since it's not used anyway), so it would be needed to re-implement `net_recv_data` function somewhere in the Zigbee area (same as some samples do in upstream Zephyr). I'll leave it up to you to decide if it's a viable solution for Zigbee.

maciekfabia replied:
Re-implementing `net_recv_data` seems to be a cleaner solution, and PR adding probably Zigbee L2 could be closed.
However, considering solutions that involve parallel Thread and Zigbee, it would be better not to implement own `net_recv_data` but to use a common one instead, so that packets could be multiplexed in layers above `net_recv_data`.

maciekfabia commented on subsys/zigbee/osif/zb_nrf_transceiver.c:
```diff
@@ -5,121 +5,143 @@
-} rx_frame_t;
+struct rx_fifo_item {
+	void *fifo_reserved;
+	struct net_pkt *pkt;
```
Great! It simplifies receiving a packet from 802.15.4 driver and passing it to ZBOSS significantly. Fixed

maciekfabia commented on subsys/zigbee/osif/zb_nrf_transceiver.c:
```diff
@@ -5,121 +5,143 @@
- */
-static rx_frame_t rx_frames[NRF_802154_RX_BUFFERS + 1];
+/* Array of received frames */
+static struct rx_fifo_item rx_fifo_buf[RX_FIFO_SIZE];
```
done

maciekfabia commented on subsys/zigbee/osif/zb_nrf_transceiver.c:
```diff
@@ -151,63 +176,93 @@ void zb_trans_get_rssi(zb_uint8_t *rssi_value_p)
 /* Configures the device as the PAN coordinator. */
 void zb_trans_set_pan_coord(zb_bool_t enabled)
 {
+	struct ieee802154_config config = { .pan_coordinator = (bool)enabled };
```
No warnings without casting. Done

greg-fer reviewed (approved):
Doc changes LGTM.

maciekfabia commented:
@rlubos I addressed all comments, could you review?
I force-pushed twice to make my fixes visilbe. First push are just my fixes, second one is rebase.

tomchy commented on subsys/zigbee/osif/zb_nrf_transceiver.c:
```diff
@@ -228,26 +275,33 @@ zb_bool_t zb_trans_is_active(void)
+				    &frag);
 		break;
 
 #ifdef ZB_ENABLE_ZGP_DIRECT
```
If we want to certify this code - the ZGP case must not be left unimplemented.

tomchy replied:
Update: agreed to add it in a separate PR.

maciekfabia replied:
I'll create a new PR for it

tomchy commented on subsys/zigbee/osif/zb_nrf_transceiver.c:
```diff
@@ -257,332 +311,234 @@ zb_bool_t zb_trans_transmit(zb_uint8_t wait_type, zb_time_t tx_at,
+		return ZB_FALSE;
+	}
+
+	state_cache.radio_state = RADIO_802154_STATE_RECEIVE;
```
I'm not an expert here, but I feel that we have a bug here.
If we start the transmission directly from the SLEEP state, does the radio peripheral go to SLEEP or RECEIVE state after transmitting?
I really miss those state diagrams from radio driver API documentation.

maciekfabia replied:
I'll try to get in debugging the actual state from radio driver. I also wrote to @hubertmis and asked for review of state caching.

hubertmis replied:
After TX radio driver automatically enters RX

hubertmis replied:
Does `radio_api->tx()` block until frame is transmitted?

tomchy commented on subsys/zigbee/osif/zb_nrf_transceiver.c:
```diff
@@ -151,63 +166,95 @@ void zb_trans_get_rssi(zb_uint8_t *rssi_value_p)
 	LOG_DBG("Function: %s", __func__);
-	(void)nrf_802154_receive();
+
+	radio_api->start(radio_dev);
```
Jest checking: is it correct that if `radio_api->start` is called during transmission, it will wait until the transmission is finished, and center the RECEIVE state afterwards?
If not, than we have some state conflicts in TX/RX functions.

maciekfabia replied:
`radio_api->start(radio_dev);` calls `nrf_802154_receive()` internally. I'm not sure if `nrf_802154_receive()` waits for completion of transmit and enters RECEIVE afterwards. Why do we need this? I mean where a state conflict can happen?

hubertmis replied:
If `nrf_802154_receive()` is called during transmission, it aborts the transmission and enters RX state.

tomchy replied:
There was a chance that the `start` operation will simply fail if called during transmission. In this case the state would be updated to RX, but the radio would be in TX.

tomchy commented on subsys/zigbee/osif/zb_nrf_transceiver.c:
```diff
@@ -257,332 +311,234 @@ zb_bool_t zb_trans_transmit(zb_uint8_t wait_type, zb_time_t tx_at,
-	ARG_UNUSED(error);
-	energy_detect.failed = true;
-	k_sem_give(&energy_detect.sem);
+	return NET_L2_MULTICAST;
```
What does it mean in Zigbee?

maciekfabia replied:
I'll change it to  `return 0`.  I forgot about it after our last talk.

maciekfabia replied:
fixed

tomchy commented on subsys/zigbee/osif/zb_nrf_transceiver.c:
```diff
@@ -257,332 +311,234 @@ zb_bool_t zb_trans_transmit(zb_uint8_t wait_type, zb_time_t tx_at,
-		break;
-	}
-#endif /* defined CONFIG_RADIO_STATISTICS */
+	return NET_OK;
```
Do I properly understand that the `CONFIG_RADIO_STATISTICS` functionality will be dropped by this PR?

maciekfabia replied:
We have some limited access to tx error after my PR https://github.com/nrfconnect/sdk-zephyr/pull/357. When it comes to rx errors, there's currently no way to access rx error code as even shim driver doesn't save it in radio driver callback.
Because of this I think collecting radio statistics has to be dropped.
I'll remove `CONFIG_RADIO_STATISTICS` from Kconfig if you don't have objections on this.

maciekfabia replied:
removed

rlubos reviewed (approved):
I don't have any more comments.

tomchy reviewed (changes requested):
* Please confirm the state caching with radio driver & shim layer experts.
* I think that we need to extend the implementation by the ZGPD transmission support.
* I do not have a strong opinion on radio statistics, but if you drop it, please align Kconfig entries too.

hubertmis reviewed (commented):
Radio state cache looks correct as long as functions cannot preempt each other. For example, when `zb_trans_enter_receive()` would preempt `zb_trans_enter_sleep()`, cache could contains `RADIO_802154_STATE_SLEEP`, while real state would be RX.

maciekfabia commented:
> Does radio_api->tx() block until frame is transmitted?

Yes, it blocks until transmission is finished and ACK frame received. In case no ACK is expected, it just blocks until transmission is finished

tomchy commented on subsys/zigbee/Kconfig:
```diff
@@ -266,12 +260,6 @@ config ZIGBEE_SHELL_CTX_MGR_ENTRIES_NBR
 
 endif #ZIGBEE_SHELL
 
-config RADIO_STATISTICS
```
Please, do not forget to make a PR changing the `zb_osif_platgform.h` file to `platform_ncs` in ZOI  :slightly_smiling_face:

tomchy replied:
Bump.

maciekfabia commented:
> Radio state cache looks correct as long as functions cannot preempt each other. For example, when `zb_trans_enter_receive()` would preempt `zb_trans_enter_sleep()`, cache could contains `RADIO_802154_STATE_SLEEP`, while real state would be RX.

We checked that with @tomchy. ZBOSS functions that change radio state, don't preempt each other.

rlubos commented on west.yml:
```diff
@@ -46,7 +46,7 @@ manifest:
     - name: zephyr
       repo-path: sdk-zephyr
-      revision: 7d20f2ebf25991b2897b91275939f8d16d38513a
+      revision: pull/357/head
```
Please update SHA: `247f9af846a29c4ffcefccb24b2a2348a2e9257f`

maciekfabia commented:
@rlubos CI passed after SHA update. I think we can merge this one too.

maciekfabia commented:
@rlubos 
We have found the cause of OTA issue and added commits that solve it.
The PR is ready to merge
